sage.py

```diff
@@ -127,29 +128,109 @@ def _is_bom_scan(scan_obj):
+        return self._get_all_items(url, 100, headers, progress_info=True)
+
+    def _get_all_project_versions(self, project_id):
+        url = self.hub.get_urlbase() + f"/api/projects/{project_id}/versions"
+        headers = {'accept': "application/vnd.blackducksoftware.project-detail-5+json"}
+        return self._get_all_items(url, 100, headers)
+
+    def _get_all_project_version_codelocations(self, project_id, version_id):
+        url = self.hub.get_urlbase() + f"/api/projects/{project_id}/versions/{version_id}/codelocations"
+        # note using key 'accept' does not work with 2020.12
+        headers = {'content-type': "application/vnd.blackducksoftware.scan-4+json"}
+        return self._get_all_items(url, 100, headers)
+
+    def _get_all_codelocations(self):
+        url = self.hub.get_urlbase() + "/api/codelocations"
+        headers = {'accept': "application/vnd.blackducksoftware.scan-4+json"}
+        return self._get_all_items(url, 100, headers, progress_info=True)
+
+    def _get_all_policies(self):
+        url = self.hub.get_urlbase() + "/api/policy-rules"
+        # note using key 'content-type' does not work with 2020.12
+        headers = {'accept': "application/vnd.blackducksoftware.policy-5+json"}
+        return self._get_all_items(url, 100, headers, progress_info=True)
+
+    def _get_all_codelocation_summaries(self, codelocation_id):
+        url = self.hub.get_urlbase() + f"/api/codelocations/{codelocation_id}/scan-summaries"
+        headers = {'accept': "application/vnd.blackducksoftware.scan-4+json"}
+        return self._get_all_items(url, 100, headers)
+
+    def _get_all_job_statistics(self):
+        # This endpoint is not in the REST API docs with 2021.2 but it still works
+        url = self.hub.get_urlbase() + "/api/job-statistics"
+        headers = {'accept': "application/vnd.blackducksoftware.status-4+json"}
+        return self._get_all_items(url, 100, headers, progress_info=True)
+
     def _get_data(self):
         last_authentication = datetime.now()
 
         '''Retrieve all the projects, versions, and scans and put them into self.data for
         subsequent analysis.
         '''
-        logging.debug('Retrieving projects')
-        projects = self.hub.get_projects(limit=99999).get('items', [])
+        logging.info("Fetching projects...")
+        projects = self._get_all_projects()
+        logging.info("Fetched %i projects", len(projects))
         total_versions = 0
+        project_count = 0
         for project in projects:
+            project_count += 1
+            m = re.match(r".*/projects/(.*)", project['_meta']['href'])
+            project_id = m.group(1)
             project_name = project['name']
-            logging.debug("Retrieving versions for project {}".format(project_name))
+            print("Project ({}/{}): {};  versions:".format(project_count, len(projects), project_name), end='', flush=True)
```

Comment:
Why use **print** instead of the logging framework? Using the logging framework allows for more flexibly routing the output where you want it to go? e.g. stdout, stderr, a file? independent of the code.

Did you consider using the newer formatting, e.g. f"Project ({project_count}/len(projects)}: {project_name}; versions:" 
Tends to make the code cleaner.

Reply:
I use print/stdout mainly to be able to print a partial string (without a newline), pause while waiting for a call to return, and then displaying the count. When observing the output running real-time, it is possible to visibly detect longer running calls. E.g. if it pauses for several seconds here:
Project (1/3): My Project;  versions:
We know that it is waiting on the get versions for the project in question.

Thank you for the tip for the newer formatting f"...{}" I've been using a mix of it with the more traditional means depending on what is more readable.